ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -156,100 +156,206 @@ data ValueHandle m = ValueHandle
+  fmap (fmap (deserialiseBS "a value" decVal)) $ getBS db (toStrictByteString (encKey k))
+
+withSerializationPut ::
+     CodecMK k v
+  -> LMDB.Database k v
+  -> k
+  -> Maybe v
   -> LMDB.Transaction LMDB.ReadWrite ()
-lmdbWriteTable db0 (HD.UtxoDiff m) =
-    void $ Map.traverseWithKey go m
+withSerializationPut (CodecMK encKey encVal _ _) db k =
+    maybe (void $ LMDB.Internal.deleteBS db keyBS) (putBS db keyBS . toStrictByteString . encVal)
   where
-    db = coerceDatabase db0
-    go k (HD.UtxoEntryDiff v reason) =
-      LMDB.put db (LmdbBox k) (if reason `elem` [HD.UedsDel, HD.UedsInsAndDel] then Nothing else Just (LmdbBox v))
+    keyBS = toStrictByteString (encKey k)
 
-lmdbInitRangeReadTable :: LedgerConstraint k v
-  => Int
+withSerializationFoldrWithKey ::
+     (k -> v -> b -> b)
+  -> b
+  -> CodecMK k v
   -> LMDB.Database k v
-  -> LMDB.Transaction mode (HD.UtxoValues k v)
-lmdbInitRangeReadTable count db0 =
-  -- This is inadequate. We are folding over the whole table, the
-  -- fiddling with either is to short circuit as best we can.
-  -- TODO improve lmdb-simple bindings to give better access to cursors
-    wrangle <$> LMDB.foldrWithKey go (Right Map.empty) db
+  -> LMDB.Transaction mode b
+withSerializationFoldrWithKey f b (CodecMK _ _ decKey decVal) = foldrWithKey fBS b
   where
-    wrangle = either HD.UtxoValues HD.UtxoValues
-    db = coerceDatabase db0
-    go (LmdbBox k) (LmdbBox v) acc = do
-      m <- acc
-      when (Map.size m >= count) $ Left m
-      pure $ Map.insert k v m
-
-lmdbRangeReadTable :: LedgerConstraint k v
-  => Int
-  -> LMDB.Database k v
-  -> HD.UtxoKeys k v
-  -> LMDB.Transaction mode (HD.UtxoValues k v)
-lmdbRangeReadTable count db0 (HD.UtxoKeys keys) = case Set.lookupMax keys of
-  -- This is inadequate. We are folding over the whole table, the
-  -- fiddling with either is to short circuit as best we can.
-  -- TODO improve lmdb-simple bindings to give better access to cursors
-  Nothing -> pure $ HD.UtxoValues Map.empty
-  Just last_excluded_key ->
-    let
-      db      = coerceDatabase db0
-      wrangle = either HD.UtxoValues HD.UtxoValues
-      go (LmdbBox k) (LmdbBox v) acc
-        | k <= last_excluded_key = acc
-        | otherwise = do
-            m <- acc
-            when (Map.size m >= count) $ Left m
-            pure $ Map.insert k v m
-    in
-      wrangle <$> LMDB.foldrWithKey go (Right Map.empty) db
+    fBS keyBS valBS =
+      f (deserialiseBS "a key" decKey keyBS) (deserialiseBS "a value" decVal valBS)
+
+-- | Deserialise a 'BS.ByteString' using the provided decoder.
+deserialiseBS
+  :: String
+  -- ^ Label to be used for error reporting. This should describe the value to be deserialised.
+  -> (forall s . CBOR.Decoder s a)
+  -> BS.ByteString
+  -> a
+deserialiseBS label decoder bs = either err snd $ deserialiseFromBytes decoder $ LBS.fromStrict bs
+  where
+    err = error $ "withSerializationFoldrWithKey: error deserializing " ++ label ++ " from the database."
```

Comment:
Would we want to be able to catch this error outside of IO or is it okay to use `error`?

Reply:
I guess this would have to be decided in CAD-4307